httpreplay/cobweb.py

```diff
@@ -175,3 +180,245 @@ def handle(self, s, ts, protocol, sent, recv):
+
+        # Contains the functions to be called when this command is the first
+        # string in a request message
+        self._commands = {
+            "ehlo": self.handle_hostname,
+            "helo": self.handle_hostname,
+            "mail": self.handle_mail,
+            "rcpt": self.handle_rcpt,
+            "auth": self.handle_auth,
+            "etrn": None, "turn": None, "atrn": None, "size": None,
+            "etrn": None, "pipelining": None, "chunking": None, "data":None,
+            "dsn": None, "rset": None, "vrfy": None, "help": None,
+            "quit": None, "noop": None, "expn": None, "binarymime": None,
+            "relay": None, "size": None, "starttls": None, "checkpoint": None,
+            "enhancedstatuscodes": None, "8bitmime": None, "send": None
+        }
+
+        # Contains the functions to be called
+        # in the request handler after the server sents this response code
+        self._res_codes = {
+            334: self.handle_auth_serv_response,
+            354: self.handle_mailbody
+        }
+
+    def handle(self, s, ts, protocol, sent, recv):
+        if protocol != "tcp":
+            self.parent.handle(s, ts, protocol, sent, recv)
+            return
+
+        if self.stream is None:
+            self.stream = self.parent.tcp.streams[s]
+
+        self.parse_request(sent)
+        self.parse_reply(recv)
+
+        if self.stream.state in ["conn_finish", "conn_closed"]:
+            self.parent.handle(s, ts, "smtp", self.request, self.reply)
+
+    def handle_hostname(self, data):
+        if len(data) > 1:
+            self.request.hostname = data[1]
+
+    def handle_rcpt(self, data):
+        if len(data) < 2:
+            return
+        for val in data:
+            self.request.mail_to.extend(re.findall('<(.*?)>', val, re.DOTALL))
+
+    def handle_mail(self, data):
+        if len(data) < 2:
+            return
+        for val in data:
+            self.request.mail_from.extend(re.findall('<(.*?)>', val, re.DOTALL))
+
+    def handle_mailbody(self, data):
+        headers_mes = data.split("\r\n\r", 1)
+        if len(headers_mes) < 2:
+            return
+
+        self.request.message = headers_mes[1]
+
+        for header in headers_mes[0].split("\r\n"):
+            if ":" not in header:
+                continue
+
+            key, value = header.split(":", 1)
+            self.request.headers[key] = value
+
+    def handle_auth(self, data):
+        """
+        Determines what kind of authentication type was used
+        and tries to collect the credentials based on the used authentication
+        type
+
+        http://www.samlogic.net/articles/smtp-commands-reference-auth.htm
+        """
+        auth_handlers = {
+            "plain": self.handle_auth_plain,
+            "login": self.handle_auth_login
+        }
+
+        if len(data) < 2:
+            return
+
+        arg_first = data[1].lower()
+        if arg_first not in auth_handlers:
+            return
```

Comment:
Now this could be a place where you do want to add logging (or even an exception, provided that we don't expect to run into new protocols anytime soon..)